Orange/widgets/unsupervised/owdistances.py

```diff
@@ -157,6 +160,9 @@ def _fix_missing():
                       _fix_discrete, _fix_missing, _fix_nonbinary):
             if not check():
                 return None
+        if (METRICS[self.metric_idx][0] == 'Bhattacharyya') and _min(data.X) < 0:
```

Comment:
It is in general a bad idea to use string literals like this. Somebody will rename it ... and there we go. I'd be happier with `METRICS[self.metric_idx][1] is distance.Bhattacharyya` (and, by the way, you can remove the parentheses).

On the other hand, why wouldn't the distance itself (i.e. function `distance. _bhattacharyya`) test the values and raise `ValueError("Bhattcharyya distance requires non-negative values")`? The widget already catches and shows `ValueError` exceptions.

As I proposed on Friday, this is better because it gives a reasonable error also to anybody that would call this distance from a script. With current code, if somebody calls Bhattacharyya with negative values, (s)he will get just a `RuntimeWarning: invalid value encountered in sqrt`, the result will be `nan` ... and this `nan` will propagate until an unrelated function crashes down the road. Testing and raising an exception there is better because it helps debugging.

This is open for discussion -- on Friday.

Reply:
I don't think it's really open for discussion, since you are obviously right. Thanks for being courteous about it.